<?xml version="1.0" encoding="UTF-8"?>
<!-- Isaac Sim launch file for humanoid training -->
<launch>
  <!-- Arguments -->
  <arg name="headless_mode" default="True" description="Run in headless mode for training"/>
  <arg name="enable_cameras" default="False" description="Enable camera sensors (disable for faster training)"/>
  <arg name="enable_physics" default="True" description="Enable physics simulation"/>
  <arg name="robot_model_path" default="$(find-pkg-share robot_description)/urdf/humanoid.urdf.xacro" description="Path to robot URDF"/>
  <arg name="training_mode" default="True" description="Optimize for training performance"/>
  
  <!-- Isaac Sim application -->
  <executable 
    cmd="python -m omni.isaac.sim.python.simulator --no-window --no-display" 
    name="isaac_sim" 
    output="screen"
    condition="$(eval arg('headless_mode'))">
  </executable>
  
  <executable 
    cmd="python -m omni.isaac.sim.python.simulator" 
    name="isaac_sim" 
    output="screen" 
    condition="$(eval not arg('headless_mode'))">
  </executable>
  
  <!-- Robot spawning node -->
  <node 
    pkg="isaac_ros_launch" 
    exec="spawn_robot.py" 
    name="robot_spawner" 
    output="screen"
    args="--robot-urdf-path $(find-pkg-share robot_description)/humanoid.urdf.xacro
          --robot-name humanoid_robot
          --position 0 0 1.0
          --orientation 0 0 0 1">
    <param name="robot_description" value="$(find-pkg-share robot_description)/humanoid.urdf.xacro"/>
    <param name="physics_dt" value="0.001"/>
    <param name="rendering_dt" value="0.01667"/>
  </node>
  
  <!-- Isaac ROS bridge -->
  <node 
    pkg="isaac_ros_bridge" 
    exec="ros_bridge_node" 
    name="isaac_ros_bridge" 
    output="screen">
    <param name="use_sim_time" value="True"/>
  </node>
  
  <!-- RL training node -->
  <node 
    pkg="humanoid_rl" 
    exec="train_humanoid_rl.py" 
    name="rl_trainer" 
    output="screen">
    <param name="max_episodes" value="5000"/>
    <param name="episode_length" value="1000"/>
    <param name="save_frequency" value="100"/>
  </node>
  
  <!-- vSLAM node (for navigation) -->
  <node 
    pkg="vslam_system" 
    exec="vslam_node" 
    name="vslam_node" 
    output="screen"
    unless="$(eval not arg('enable_cameras'))">
    <param name="camera_topic" value="/humanoid_robot/head_camera/image_raw"/>
    <param name="imu_topic" value="/humanoid_robot/imu/data"/>
  </node>
  
  <!-- Curriculum learning manager -->
  <node 
    pkg="curriculum_learning" 
    exec="curriculum_manager.py" 
    name="curriculum_manager" 
    output="screen">
    <param name="config_path" value="$(find-pkg-share configs)/humanoid_training_config.json"/>
  </node>
  
  <!-- Navigation stack -->
  <include file="$(find-pkg-share nav2_bringup)/launch/navigation_launch.py" 
           unless="$(eval not arg('training_mode'))"/>
  
  <!-- RViz for visualization (only enabled when not in training mode) -->
  <node 
    pkg="rviz2" 
    exec="rviz2" 
    name="rviz2" 
    args="-d $(find-pkg-share physical_ai_book)/rviz/humanoid_navigation.rviz"
    if="$(eval not arg('training_mode'))">
  </node>
  
  <!-- Training monitoring -->
  <node 
    pkg="tensorboard_logger" 
    exec="tensorboard_node" 
    name="tensorboard" 
    output="screen">
    <param name="log_dir" value="./tb_logs/"/>
  </node>
</launch>